File: go/InsertIntervals/InsertIntervals.go
package main

import "fmt"

// Time -> 0(N) Space -> 0(N)
func insertIntervals(intervals [][]int, newInterval []int) [][]int {
	if len(intervals) == 0 {
		return [][]int{newInterval}
	}
	insertedInterval := make([][]int, 0)

	for idx := range intervals {
		if intervals[idx][0] > newInterval[0] {
			insertedInterval = append(insertedInterval, newInterval)
			insertedInterval = append(insertedInterval, intervals[idx:]...)
			break
		}
		insertedInterval = append(insertedInterval, intervals[idx])
	}

	if len(intervals) == len(insertedInterval) {
		insertedInterval = append(insertedInterval, newInterval)
	}

	mergedInterval := make([][]int, 0)

	i := 0
	for i < len(insertedInterval) {
		start := i
		i++
		for i < len(insertedInterval) && insertedInterval[i][0] <= insertedInterval[start][1] {
			insertedInterval[start][1] = max(insertedInterval[i][1], insertedInterval[start][1])
			i++
		}
		mergedInterval = append(mergedInterval, insertedInterval[start])
	}

	return mergedInterval
}

func main() {
	fmt.Println(insertIntervals([][]int{
		{1, 2}, {3, 5}, {6, 7}, {8, 10}, {12, 16},
	}, []int{4, 8}))
	fmt.Println(insertIntervals([][]int{
		{1, 5},
	}, []int{3, 7}))
}

File: python/insertIntervals.py
# Time -> 0(N) Space -> 0(N)
def insertIntervals(
    intervals: list[list[int]], newInterval: list[int]
) -> list[list[int]]:
    insertedIntervals: list[list[int]] = []

    for idx in range(len(intervals)):
        if intervals[idx][0] > newInterval[0]:
            insertedIntervals.append(newInterval)
            insertedIntervals.extend(intervals[idx : len(intervals)])
            break
        insertedIntervals.append(intervals[idx])

    if len(intervals) == len(insertedIntervals):
        insertedIntervals.append(newInterval)

    mergedIntervals: list[list[int]] = []

    i = 0
    while i < len(insertedIntervals):
        start = i
        i += 1
        while (
            i < len(insertedIntervals)
            and insertedIntervals[i][0] <= insertedIntervals[start][1]
        ):
            insertedIntervals[start][1] = max(
                insertedIntervals[start][1], insertedIntervals[i][1]
            )
            i += 1

        mergedIntervals.append(insertedIntervals[start])
    return mergedIntervals


print(insertIntervals([[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [4, 8]))
